src/main/java/io/jenkins/plugins/sprp/PipelineSnippetGenerator.java

```diff
@@ -99,19 +105,134 @@ else if(agent.getDockerfile() != null){
 
                 agentLines.add(getCommonOptionsOfAgent(agent));
                 agentLines.add("}");
+                agentLines.add("}");
             }
-            else {
+            else if(agent.getLabel() != null && agent.getCustomWorkspace() != null){
+                agentLines.add("agent {");
                 agentLines.add("node{");
                 agentLines.add(getCommonOptionsOfAgent(agent));
                 agentLines.add("}");
+                agentLines.add("}");
+            }
+            else {
+                agentLines.add("agent any");
             }
 
-            agentLines.add("}");
         }
 
+        if(agent != null)
+            agentLines.addAll(getTools(agent.getTools()));
+
         return agentLines;
     }
 
+    public List<String> getTools(HashMap<String, String> tools){
+        ArrayList<String> snippetLines = new ArrayList<>();
+
+        if(tools == null)
+            return snippetLines;
+
+        snippetLines.add("tools {");
+
+        for(Map.Entry<String, String> entry: tools.entrySet()){
+            snippetLines.add(entry.getKey() + " '" + entry.getValue() + "'");
+        }
+
+        snippetLines.add("}");
+
+        return snippetLines;
+    }
+
+    public List<String> getEnvironment(Environment environment){
+        ArrayList<String> snippetLines = new ArrayList<>();
+
+        if(environment == null || (environment.getVariables() == null && environment.getCredentials() == null))
+            return snippetLines;
+
+        snippetLines.add("environment {");
+
+        for(Map.Entry<String, String> entry: environment.getVariables().entrySet()){
+            snippetLines.add(entry.getKey() + " = '" + entry.getValue() + "'");
+        }
+
+        for(Credential credential: environment.getCredentials()){
+            snippetLines.add(credential.getVariable() + " = credentials('" + credential.getCredentialId() + "')");
+        }
+
+        snippetLines.add("}");
+
+        return snippetLines;
+    }
+
+    public List<String> getPostSection(Post postSection) throws NoSuchMethodException, ConfiguratorException,
+            InstantiationException, IllegalAccessException, InvocationTargetException, NoSuchFieldException {
+        ArrayList<String> snippetLines = new ArrayList<>();
+
+        if(postSection == null) {
+            return snippetLines;
+        }
+
+        snippetLines.add("post {");
+
+        if(postSection.getAborted() != null){
```

Comment:
General comment about coding style: put spaces between keywords like "if" and the opening parenthesis. This helps the reviewer quickly recognise methods from keywords. Also, put a space between the closing parenthesis `)` and the following curly brace `{`.

Reply:
I will do this in another pull request. Because I want to make this consistent in all the files, and doing it in this pull request will make review process difficult. Will it be ok @martinda .

Reply:
Yes, you can have a PR solely for the code style changes.